File: 4-digit-7-segment-display/python/4-digit-7-segment-display.py
# Copyright (c) 2025, BlackBerry Limited. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rpi_gpio as GPIO  # Import the QNX Raspberry Pi GPIO module for controlling GPIO pins
import time  # Import the time module for delays

# Define GPIO pins for each segment
segments = {
    "A": 23, "B": 6, "C": 20, "D": 5,
    "E": 24, "F": 19, "G": 12, "DP": 21
}

# Define GPIO pins for digit control (common anode configuration)
digits_pins = [18, 13, 26, 25]

# Segment configurations for displaying digits 0-9 on a 7-segment display
# Each list represents the states of segments [A, B, C, D, E, F, G, DP] (1 = ON, 0 = OFF)
digits = {
    0: [1,1,1,1,1,1,0,0], 
    1: [0,1,1,0,0,0,0,0], 
    2: [1,1,0,1,1,0,1,0], 
    3: [1,1,1,1,0,0,1,0],
    4: [0,1,1,0,0,1,1,0], 
    5: [1,0,1,1,0,1,1,0], 
    6: [1,0,1,1,1,1,1,0], 
    7: [1,1,1,0,0,0,0,0],
    8: [1,1,1,1,1,1,1,0], 
    9: [1,1,1,1,0,1,1,0]
}

# Use Broadcom (BCM) numbering scheme for GPIO pins
GPIO.setmode(GPIO.BCM)

# Set all segment and digit control pins as output
for pin in segments.values():
    GPIO.setup(pin, GPIO.OUT)

for pin in digits_pins:
    GPIO.setup(pin, GPIO.OUT)
    GPIO.output(pin, 1)  # Turn all digits OFF initially (common anode HIGH = OFF)

def display_number(num):
    num_str = str(num).zfill(4)  # Convert to 4-digit string, padding with zeros if needed
    
    for i, digit_char in enumerate(num_str):
        digit = int(digit_char)  # Convert character to integer

        GPIO.output(digits_pins[i], 0)  # Turn ON the current digit

        pattern = digits[digit]  # Get segment pattern
        for j, segment in enumerate(segments.values()):
            GPIO.output(segment, pattern[j])  # Set segment states

        time.sleep(0.001)  # Small delay
        GPIO.output(digits_pins[i], 1)  # Turn OFF the digit

try:
    while True:
        for num in range(10000):  # Cycle through 0000-9999
            for _ in range(10):  # Multiplex each number for stability
                display_number(num)

finally:
    GPIO.cleanup()  # Cleanup on exit

File: 4-digit-7-segment-display/c/src/4-digit-7-segment-display.c
/*
* Copyright (c) 2025, BlackBerry Limited. All rights reserved.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

#include <stdio.h>    // Standard input/output functions (for puts and perror)
#include <stdlib.h>   // Standard library (for EXIT_SUCCESS and EXIT_FAILURE) 

#include "rpi_gpio.h" // Raspberry Pi GPIO control library

// Define GPIO pins for each segment of the 7-segment display
#define SEG_A GPIO23
#define SEG_B GPIO6
#define SEG_C GPIO20
#define SEG_D GPIO5
#define SEG_E GPIO24
#define SEG_F GPIO19
#define SEG_G GPIO12
#define SEG_DP GPIO21 // Decimal point segment

// Define GPIO pins for digit control (for a 4-digit display)
#define DIGIT_1 GPIO18
#define DIGIT_2 GPIO13
#define DIGIT_3 GPIO26
#define DIGIT_4 GPIO25

// Representation of digits (0-9) on the 7-segment display
// Each row corresponds to a digit, and each column represents a segment (A-G, DP)
int digits[10][8] = {
    {1, 1, 1, 1, 1, 1, 0, 0},  // 0
    {0, 1, 1, 0, 0, 0, 0, 0},  // 1
    {1, 1, 0, 1, 1, 0, 1, 0},  // 2
    {1, 1, 1, 1, 0, 0, 1, 0},  // 3
    {0, 1, 1, 0, 0, 1, 1, 0},  // 4
    {1, 0, 1, 1, 0, 1, 1, 0},  // 5
    {1, 0, 1, 1, 1, 1, 1, 0},  // 6
    {1, 1, 1, 0, 0, 0, 0, 0},  // 7
    {1, 1, 1, 1, 1, 1, 1, 0},  // 8
    {1, 1, 1, 1, 0, 1, 1, 0}   // 9
};

// Array of digit control pins (common anode display)
int digit_pins[4] = {DIGIT_1, DIGIT_2, DIGIT_3, DIGIT_4};

// Initialize a given GPIO pin as an output
int init_gpio(int gpio_pin) {
    if (rpi_gpio_setup(gpio_pin, GPIO_OUT)) {
        perror("rpi_gpio_setup");
        return -1;
    }
    return 0;
}

// Turn off all digit control lines (disable all digits)
void clear_digits() {
    for (int i = 0; i < 4; i++) {
        rpi_gpio_output(digit_pins[i], GPIO_HIGH);  // Common anode: HIGH = OFF
    }
}

// Display a single digit on a specified position
void display_digit(int digit, int position) {
    clear_digits(); // Turn off all digits before enabling a new one
    
    // Get the segment pattern for the given digit
    int *pattern = digits[digit];
    
    // Set the segment states based on the pattern
    rpi_gpio_output(SEG_A, pattern[0] ? GPIO_HIGH : GPIO_LOW);
    rpi_gpio_output(SEG_B, pattern[1] ? GPIO_HIGH : GPIO_LOW);
    rpi_gpio_output(SEG_C, pattern[2] ? GPIO_HIGH : GPIO_LOW);
    rpi_gpio_output(SEG_D, pattern[3] ? GPIO_HIGH : GPIO_LOW);
    rpi_gpio_output(SEG_E, pattern[4] ? GPIO_HIGH : GPIO_LOW);
    rpi_gpio_output(SEG_F, pattern[5] ? GPIO_HIGH : GPIO_LOW);
    rpi_gpio_output(SEG_G, pattern[6] ? GPIO_HIGH : GPIO_LOW);
    rpi_gpio_output(SEG_DP, pattern[7] ? GPIO_HIGH : GPIO_LOW);
    
    // Enable the specific digit position (LOW = ON for common anode)
    rpi_gpio_output(digit_pins[position], GPIO_LOW);
    
    usleep(1000); // Small delay to allow persistence of display
}

// Display a 4-digit number on the 7-segment display
void display_number(int num) {
    char num_str[5];
    snprintf(num_str, sizeof(num_str), "%04d", num); // Format number as 4-digit string
    
    for (int i = 0; i < 4; i++) {
        int digit = num_str[i] - '0'; // Convert character to integer
        display_digit(digit, i);
    }
}

int main() {
    // Initialize all GPIO pins used for the segments and digit control
    int all_pins[] = {SEG_A, SEG_B, SEG_C, SEG_D, SEG_E, SEG_F, SEG_G, SEG_DP, 
                       DIGIT_1, DIGIT_2, DIGIT_3, DIGIT_4};
    
    for (int i = 0; i < 12; i++) {
        if (init_gpio(all_pins[i]) == -1) {
            return EXIT_FAILURE; // Exit if any GPIO setup fails
        }
    }
    
    // Continuously display numbers from 0000 to 9999
    while (1) {
        for (int num = 0; num < 10000; num++) { // Loop through numbers 0-9999
            for (int i = 0; i < 10; i++) { // Multiplex each number for stability
                display_number(num);
            }
        }
    }
    
    return EXIT_SUCCESS;
}